nexus/src/db/subnet_allocation.rs

```diff
@@ -490,142 +305,585 @@ impl QueryFragment<Pg> for FilterConflictingVpcSubnetRangesQueryValues {
     }
 }
 
+/// Errors related to inserting or attaching a NetworkInterface
+#[derive(Debug)]
+pub enum NetworkInterfaceError {
+    /// The instance specified at create or attach is associated with a
+    /// different VPC
+    InstanceSpansMultipleVpcs(Uuid),
+    /// There are no available IP addresses in the requested subnet
+    NoAvailableIpAddresses,
+    /// An explicitly-requested IP address is already in use
+    IpAddressNotAvailable(std::net::IpAddr),
+    /// Any other error
+    External(external::Error),
+}
+
+impl NetworkInterfaceError {
+    /// Construct a `NetworkInterfaceError` from a database error
+    ///
+    /// This catches the various errors that the `InsertNetworkInterfaceQuery`
+    /// can generate, especially the intentional errors that indicate either IP
+    /// address exhaustion or an attempt to attach an interface to an instance
+    /// that is already associated with another VPC.
+    pub fn from_pool(
+        e: async_bb8_diesel::PoolError,
+        interface: &IncompleteNetworkInterface,
+    ) -> Self {
+        use crate::db::error;
+        use async_bb8_diesel::ConnectionError;
+        use async_bb8_diesel::PoolError;
+        use diesel::result::DatabaseErrorKind;
+        use diesel::result::Error;
+        const MULTIPLE_VPC_ERROR_MESSAGE: &str =
+            r#"could not parse "" as type uuid: uuid: incorrect UUID length: "#;
+        const IP_EXHAUSTION_ERROR_MESSAGE: &str =
+            r#"null value in column "ip" violates not-null constraint"#;
+        const IP_NOT_AVAILABLE_CONSTRAINT: &str =
+            "network_interface_subnet_id_ip_key";
+        const NAME_CONFLICT_CONSTRAINT: &str =
+            "network_interface_instance_id_name_key";
+        match e {
+            // If the address allocation subquery fails, we'll attempt to insert
+            // NULL for the `ip` column.
+            PoolError::Connection(ConnectionError::Query(
+                Error::DatabaseError(
+                    DatabaseErrorKind::NotNullViolation,
+                    ref info,
+                ),
+            )) if info.message() == IP_EXHAUSTION_ERROR_MESSAGE => {
```

Comment:
It took me a minute to appreciate that the comments here and at L358 explain why the message that we're checking against represents the condition that we're looking for.  I wonder if this would be clearer if you had one match arm for the `PoolError::Connection(ConnectionError::Query(Error::DatabaseError(...)))` case that called some other function where you could centralize this logic (and the explanations).  What do you think?

Reply:
That all sounds good to me, I'll give it a whirl.

Reply:
I tried this out. It definitely localizes the "what the shit is going on here" vibes to one function. And if you're there, you're already in deep. Will push :)